src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/GenerateMethodCodeActionProvider.cs

```diff
@@ -75,51 +82,87 @@ private static List<RazorVSInternalCodeAction> CreateCodeAction(GenerateMethodCo
+                EventName = eventName,
+                IsAsync = false,
+            },
+            new GenerateMethodCodeActionParams()
+            {
+                Uri = context.Request.TextDocument.Uri,
+                MethodName = methodName,
+                EventName = eventName,
+                IsAsync = true,
             }
+        };
+
+        return true;
+    }
 
-            methodName = content;
+    private static bool TryParseMethodAndEventName(
+        SyntaxNode owner,
+        [NotNullWhen(true)] out string? methodName,
+        [NotNullWhen(true)] out string? eventName)
+    {
+        methodName = null;
+        eventName = null;
+
+        var commonParent = owner.Kind == SyntaxKind.CSharpExpressionLiteral ? owner.Parent.Parent : owner.Parent;
+        if (commonParent.Kind != SyntaxKind.MarkupTagHelperDirectiveAttribute)
+        {
+            return false;
         }
-        else
+
+        var children = commonParent.ChildNodes();
+        var csharpExpressionLiteralNode = owner;
+
+        if (csharpExpressionLiteralNode.Kind == SyntaxKind.MarkupTextLiteral)
         {
-            var children = parent.ChildNodes();
-            foreach (var child in children)
+            // We have a MarkupTextLiteral and need to find the CSharpExpressionLiteral node.
+            // The hierarchy here starting with the parent of the MarkupTextLiteral node is:
+            // MarkupTagHelperDirectiveAttribute > MarkupTagHelperAttributeValue > CSharpExpressionLiteral
+            var markupTagHelperAttributeValueNode = children.FirstOrDefault(c => c.Kind == SyntaxKind.MarkupTagHelperAttributeValue);
+            csharpExpressionLiteralNode = markupTagHelperAttributeValueNode?.ChildNodes()
+                .FirstOrDefault(c => c.Kind == SyntaxKind.CSharpExpressionLiteral);
+            if (csharpExpressionLiteralNode is null)
             {
-                if (child.Kind == SyntaxKind.MarkupTagHelperAttributeValue)
-                {
-                    var content = child.GetContent();
-                    if (SyntaxFacts.IsValidIdentifier(content))
-                    {
-                        methodName = content;
-                        break;
-                    }
-                }
+                return false;
             }
         }
 
-        if (methodName.IsNullOrEmpty())
+        var content = csharpExpressionLiteralNode.GetContent();
+        if (!SyntaxFacts.IsValidIdentifier(content))
         {
             return false;
         }
 
-        @params = new GenerateMethodCodeActionParams()
+        methodName = content;
+
+        for (var i = 1; i < children.Count; i++)
         {
-            Uri = context.Request.TextDocument.Uri,
-            MethodName = methodName,
-        };
+            if (children[i].Kind == SyntaxKind.Equals)
+            {
+                // We assume that the node that contains the event name will always come before the one equals node.
```

Comment:
The code for hooking up a callback in blazor is:
`<button @onclick="Method"/>`
so the event name is always going to be before the `=`. I realize from David's comment though that we actually don't need to do this since `commonParent` actually has a property called `TagHelperAttributeInfo` that has the event name in it so this for loop here will be removed completely. 😄